python/tvm/micro/contrib/stm32/emitter.py

```diff
@@ -0,0 +1,1557 @@
+                if storage_id not in buffer_list_:
+                    buffer_list_[storage_id] = []
+                buffer_entry = buffer_list_[storage_id]
+                buffer_entry.append(tensor)
+            else:
+                #
+                # Compute the offset
+                #
+                dltype = tensor["dltype"]
+                aligned_offset = _get_aligned_offset(offset, dltype)
+                tensor["offset"] = aligned_offset
+
+            self.input_data_[dl_tensor_name] = tensor
+
+            #
+            # Next offset
+            #
+            offset = aligned_offset + tensor["size"]
+
+        #
+        # Compute 'output_data_'
+        #
+        offset = 0
+        for output in self.outputs_:
+            nid = output[0]
+            idx = output[1]
+
+            node = self.nodes_[nid]
+            node_name = node["name"]
+
+            dl_tensor_name = get_output_tensor_name(node_name, idx)
+
+            tensor = self.__get_tensor_from_node(nid, idx)
+
+            if self.outputs_static == True:
+                #
+                # Remember this tensor with the storage id
+                #
+                storage_id = tensor["storage_id"]
+                if storage_id not in buffer_list_:
+                    buffer_list_[storage_id] = []
+                buffer_entry = buffer_list_[storage_id]
+                buffer_entry.append(tensor)
+            else:
+                #
+                # Compute the offset
+                #
+                dltype = tensor["dltype"]
+                aligned_offset = _get_aligned_offset(offset, dltype)
+                tensor["offset"] = aligned_offset
+
+            self.output_data_[dl_tensor_name] = tensor
+
+            #
+            # Next offset
+            #
+            offset = aligned_offset + tensor["size"]
+
+        #
+        # Go over all storage IDs and compute offsets and activations_size_
+        #
+        offset = 0
+        for storage_id in buffer_list_:
+            buffer_entry = buffer_list_[storage_id]
+
+            new_offset = offset
+            for tensor in buffer_entry:
+                assert tensor["storage_id"] == storage_id
+                dltype = tensor["dltype"]
+                aligned_offset = _get_aligned_offset(offset, dltype)
+                tensor["offset"] = aligned_offset
+                size = tensor["size"]
+                if (aligned_offset + size) > new_offset:
+                    new_offset = aligned_offset + size
+            offset = new_offset
+
+        self.activations_size_ = offset
+
+    def _parse_model(self, quantization=None):
+        """Parse the module. Build internal data structures.
+
+        Parameters
+        ----------
+        module : TVM module or ModuleLibraryFormat object
+           The module to parse
+
+        quantization: Dictionary
+           The quantization information for model inputs/outputs.
+        """
+
+        for key in self.graph_:
```

Comment:
from other file: i would suggest not extracting these from self.graph_ and instead just accessing them using self.graph_["nodes"]. it may be very slightly slower, but it makes the code easier to maintain as you don't need to keep track of where each of these properties comes from. if you want to assert there aren't unknown keys in the graph, you could assert(list(sorted(self.graph_)) == ["arg_nodes", "attrs", "heads", "metadata", "node_row_ptr", "nodes"])

Reply:
I prefer to keep a separate data structure for the emitter. Generally, this data structure is not necessarily equal to the TVM graph_['nodes']. For example, I may decide to not include OPs that are a 'noop' in the code, etc. This is done on purpose to simplify development. No problem maintaining - the structure is built in a single place from the graph.

Reply:
ok, i don't think we need to block on this but i don't see you doing any modification to `self._nodes`, `self._arg_nodes`, `self._node_row_ptr`, `self._outputs`, or `self._attrs` right now.